File: 2000290120169/Binary_search.py

def binary_search(arr, l, r, x):
    if l <= r:
        mid = l+(r-l)//2

        if arr[mid] == x:
            return mid
        if x > arr[mid]:
            return binary_search(arr, mid+1, r, x)
        else:
            return binary_search(arr, l, mid-1, x)
    return -1

arr = [2,3,5,6,7,8,9] #sorted arr
x = 5
findx = binary_search(arr, 0,len(arr)-1, x)

if findx == -1:
    print(x, "Not Found !!")
else:
    print(x, "Found at index: ", findx)


#Iterative Binary Search

def binary_search(arr, l, r, x):
    while l <= r:
        mid = l+(r-l)//2

        if arr[mid] == x:
            return mid
        if x > arr[mid]:
            l = mid + 1
        else:
            r = mid-1
    return -1

arr = [2,3,5,6,7,8,9] #sorted arr
x = 5
findx = binary_search(arr, 0,len(arr)-1, x)

if findx == -1:
    print(x, "Not Found !!")
else:
    print(x, "Found at index: ", findx)

File: 2100290129008/Week 2/binary-search.cpp
#include <bits/stdc++.h>
using namespace std;

// Iterative Binary Search
int binarySearch(int arr[], int len, int find){
	int low = 0;
	int high = len - 1;

	while(low <= high){
		int mid = low + (high - low)/2;
		if(arr[mid] == find)
			return mid;
		else if(arr[mid] < find)
			low = mid + 1;
		else
			high = mid - 1;
	}
	return -1;
}


// Recursive Binary Search
int recursiveBinarySearch(int arr[], int low, int high, int find){
	int mid = low + (high - low)/2;

	while(low <= high){
		if(arr[mid] == find)
			return mid;
		else if(arr[mid] < find)
			return recursiveBinarySearch(arr,mid + 1,high,find);
		else if(arr[mid] > find)
			return recursiveBinarySearch(arr,low,mid - 1,find);
	}
	return -1;
}


int main()
{
	int arr[10] = {1,2,3,4,5,7,8,9,11,15};
	int len = 10;
	int find = 7;
	if(binarySearch(arr,len,find) == -1)
		cout << "Element not found\n";
	else
		cout << "Element found at index " << binarySearch(arr,len,find) << endl;

	cout << "Calling Recursive Binary Search Function" << endl;

	int low = 0;
	int high = len - 1;
	if(recursiveBinarySearch(arr,low,high,find) == -1)
		cout << "Element not found\n";
	else
		cout << "Element found at index " << recursiveBinarySearch(arr,low,high,find) << endl;

    return 0;
}
